lib/evmone/eof.cpp

```diff
@@ -88,83 +143,414 @@ std::pair<EOFSectionHeaders, EOFValidationError> validate_eof_headers(bytes_view
     if (state != State::terminated)
         return {{}, EOFValidationError::section_headers_not_terminated};
 
-    const auto section_bodies_size = section_headers[CODE_SECTION] + section_headers[DATA_SECTION];
+    const auto section_bodies_size =
+        (!section_headers[TYPE_SECTION].empty() ? section_headers[TYPE_SECTION].front() : 0) +
+        std::accumulate(
+            section_headers[CODE_SECTION].begin(), section_headers[CODE_SECTION].end(), 0) +
+        (!section_headers[DATA_SECTION].empty() ? section_headers[DATA_SECTION].front() : 0);
     const auto remaining_container_size = container_end - it;
     if (section_bodies_size != remaining_container_size)
         return {{}, EOFValidationError::invalid_section_bodies_size};
 
+    if (!section_headers[TYPE_SECTION].empty() &&
+        section_headers[TYPE_SECTION][0] != section_headers[CODE_SECTION].size() * 4)
+        return {{}, EOFValidationError::invalid_type_section_size};
+
     return {section_headers, EOFValidationError::success};
 }
 
+std::pair<std::vector<EOF1TypeHeader>, EOFValidationError> validate_types(
+    bytes_view container, size_t header_size, std::vector<uint16_t> type_section_sizes) noexcept
+{
+    assert(!container.empty());              // guaranteed by EOF headers validation
+    assert(type_section_sizes.size() <= 1);  // guaranteed by EOF headers validation
+
+    if (type_section_sizes.empty())
+        return {{{0, 0, 0}}, EOFValidationError::success};
+
+    std::vector<EOF1TypeHeader> types;
+
+    // guaranteed by EOF headers validation
+    assert(header_size + type_section_sizes[0] < container.size());
+
+    for (auto offset = header_size; offset < header_size + type_section_sizes[0]; offset += 4)
+    {
+        types.emplace_back(EOF1TypeHeader{container[offset], container[offset + 1],
+            static_cast<uint16_t>(container[offset + 2] << 8 | container[offset + 3])});
+    }
+
+    // check 1st section is (0, 0)
+    if (types[0].inputs_num != 0 || types[0].outputs_num != 0)
+        return {{}, EOFValidationError::invalid_first_section_type};
+
+    for (const auto& t : types)
+    {
+        if (t.max_stack_height > MAX_STACK_HEIGHT)
```

Comment:
According to EIP-5450 max stack height doesn't exceed 1023
```suggestion
        if (t.max_stack_height >= MAX_STACK_HEIGHT)
```